gdal/frmts/sentinel2/sentinel2dataset.cpp

```diff
@@ -2155,6 +2252,100 @@ static bool SENTINEL2GetGranuleList_L1CSafeCompact(CPLXMLNode* psMainMTD,
+{
+    const char* pszNodePath = "=Level-2A_User_Product.General_Info.Product_Info";
+    CPLXMLNode* psProductInfo = CPLGetXMLNode(psMainMTD, pszNodePath);
+    if( psProductInfo == nullptr )
+    {
+        pszNodePath = "=Level-2A_User_Product.General_Info.L2A_Product_Info";
+        psProductInfo = CPLGetXMLNode(psMainMTD, pszNodePath);
+    }
+    if( psProductInfo == nullptr )
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Cannot find %s",
+                        pszNodePath);
+        return false;
+    }
+
+    CPLXMLNode* psProductOrganisation =
+                        CPLGetXMLNode(psProductInfo, "Product_Organisation");
+    if( psProductOrganisation == nullptr )
+    {
+        psProductOrganisation =
+                        CPLGetXMLNode(psProductInfo, "L2A_Product_Organisation");
+        if( psProductOrganisation == nullptr )
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "Cannot find %s", "Product_Organisation");
+            return false;
+        }
+    }
+
+    CPLString osDirname( CPLGetDirname(pszFilename) );
+#ifdef HAVE_READLINK
+    char szPointerFilename[2048];
+    int nBytes = static_cast<int>(readlink(pszFilename, szPointerFilename,
+                                           sizeof(szPointerFilename)));
+    if (nBytes != -1)
+    {
+        const int nOffset =
+            std::min(nBytes, static_cast<int>(sizeof(szPointerFilename)-1));
+        szPointerFilename[nOffset] = '\0';
+        osDirname = CPLGetDirname(szPointerFilename);
+    }
+#endif
+
+    const char chSeparator = SENTINEL2GetPathSeparator(osDirname);
+    for(CPLXMLNode* psIter = psProductOrganisation->psChild; psIter != nullptr;
+                                                    psIter = psIter->psNext )
+    {
+        if( psIter->eType != CXT_Element ||
+            !EQUAL(psIter->pszValue, "Granule_List") )
+        {
+            continue;
+        }
+        for(CPLXMLNode* psIter2 = psIter->psChild; psIter2 != nullptr;
+                                                     psIter2 = psIter2->psNext )
+        {
+            if( psIter2->eType != CXT_Element ||
+                !EQUAL(psIter2->pszValue, "Granule") )
+            {
+                continue;
+            }
+
+            const char* pszImageFile = CPLGetXMLValue(psIter2, "IMAGE_FILE", nullptr);
+            if( pszImageFile == nullptr )
+            {
+                pszImageFile = CPLGetXMLValue(psIter2, "IMAGE_FILE_2A", nullptr);
+                if( pszImageFile == nullptr || strlen(pszImageFile) < 3 )
+                {
+                    CPLDebug("SENTINEL2", "Missing IMAGE_FILE element");
+                    continue;
+                }
+            }
+            L1CSafeCompatGranuleDescription oDesc;
+            oDesc.osBandPrefixPath = osDirname + chSeparator + pszImageFile;
+            oDesc.osBandPrefixPath.resize( oDesc.osBandPrefixPath.size() - 36 );
```

Comment:
oDesc.osBandPrefixPath.size() should be first compared against 36

Reply:
Size checks are implemented.